src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,242 @@
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    // set K0 to upper position (note that "upper" and "lower" refer to the schematic)
+    fn engage_k0(&mut self) {
+        let mut mcp = self.gpio.try_lock().unwrap();
+        // set flipflop data pin
+        mcp.set_gpio(self.k0_d.into(), Level::High).unwrap();
+        // set flipflop clock input low to prepare rising edge
+        mcp.set_gpio(self.k0_cp.into(), Level::Low).unwrap();
+        // set flipflop clock input high to generate rising edge
+        mcp.set_gpio(self.k0_cp.into(), Level::High).unwrap();
+    }
+
+    // set K0 to lower position
+    fn disengage_k0(&mut self) {
+        let mut mcp = self.gpio.try_lock().unwrap();
+        mcp.set_gpio(self.k0_d.into(), Level::High).unwrap();
+        mcp.set_gpio(self.k0_cp.into(), Level::Low).unwrap();
+        mcp.set_gpio(self.k0_cp.into(), Level::High).unwrap();
+    }
+
+    // set K1 to upper position
+    fn disengage_k1(&mut self) {
+        let mut mcp = self.gpio.try_lock().unwrap();
+        // set en high and en _n low in order to engage K1
+        mcp.set_gpio(self.k1_en.into(), Level::Low).unwrap();
+        mcp.set_gpio(self.k1_en_n.into(), Level::High).unwrap();
+    }
+
+    // set K1 to lower position
+    fn engage_k1(&mut self) {
+        let mut mcp = self.gpio.try_lock().unwrap();
+        mcp.set_gpio(self.k1_en.into(), Level::High).unwrap();
+        mcp.set_gpio(self.k1_en_n.into(), Level::Low).unwrap();
+    }
+}
+
+impl<I2C, E> sm::StateMachine<Relay<I2C>>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    /// Start relay enabling sequence. Returns the relay delay we need to wait for.
+    pub fn enable(&mut self) -> fugit::MillisDuration<u64> {
+        self.process_event(sm::Events::Enable).unwrap();
+        Relay::<I2C>::K0_DELAY // engage K0 first
+    }
+
+    pub fn disable(&mut self) -> fugit::MillisDuration<u64> {
+        self.process_event(sm::Events::Disable).unwrap();
+        Relay::<I2C>::K1_DELAY // engage K1 first
+    }
+
+    /// Handle a completed relay transition. Returns `Some(relay delay)` if we need to wait,
+    /// otherwise returns `None`.
+    pub fn handle_relay(&mut self) -> Option<fugit::MillisDuration<u64>> {
+        self.process_event(sm::Events::RelayDone).unwrap();
+        match *self.state() {
+            sm::States::EnableWaitK1 => Some(Relay::<I2C>::K1_DELAY), // disengage K1 second
+            sm::States::DisableWaitK0 => Some(Relay::<I2C>::K0_DELAY), // disengage K0 second
+            _ => None, // done, no delay needed
+        }
+    }
+}
+
+/// A SharedMcp can provide ownership of one of two sets of relays on Driver.
+/// Each set consists of two relays that control the state of a Driver output channel.
+/// The relays are controlled by toggeling pins on an MCP23008 I2C IO expander on the Driver board.
+/// Both sets use the same MCP23008 chip, hence the SharedMcp.
+pub struct SharedMcp<I2C1> {
```

Comment:
I would remove this struct and inline the `Relay::new(&mutex, ch)` call.
It really doesn't seem to provide any functionality or abstraction.